nixos/modules/services/editors/emacs.nix

```diff
@@ -80,6 +80,15 @@ in
         using the EDITOR environment variable.
       '';
     };
+
+    startWithGraphical = mkOption {
```

Comment:
I would rather not introduce a new top-level bool option.
My suggestion is to:

- Remove the option
- Move the behaviour to `wantedBy` (guarded by `stateVersion`)
- Add a release note documenting the breaking behaviour

Then someone who wishes for a different behaviour can directly override the system unit and we don't need another option.

Reply:
> I think this new option make sense and better than introducing breaking behavior. People who use terminal client won't need graphical session.
 
Boolean options compose extremely poorly, in the future there might be another use case "warranting" an option and then you have two mutually exclusive options.
If there was to be a top-level option it should absolutely not be a boolean.

> People who use terminal client won't need graphical session.

My suggested approach will not break anything for people who use X terminals. This change would only break things for people who:
- Have a graphical setup
- And try to launch emacsclient from a TTY without the graphical session running

This is a very small price to pay for better long term maintainability and simplicity, people who need a graphical setup & start emacsclient from a tty are the only ones who will have to override the systemd unit's `wantedBy`.

Reply:
Guarding this behind stateVersion seems odd to me. It has nothing to do with on-disk state.

This option does not need to compose. You can't both want it and not want it.  
You either get a sensible default for common use-cases or set it on your own terms.

Reply:
> Guarding this behind stateVersion seems odd to me. It has nothing to do with on-disk state.

`stateVersion` is not used only for state but also for breaking behaviour.

> This option does not need to compose. You can't both want it and not want it.

A `startWithGraphical` bool option has several issues:
- Poor composability (yes, I think this is important to think about _all_ the time in NixOS modules)
- Imprecise naming: This option doesn't actually mean "start with graphical". It means start _only_ with graphical.

Making this a string option, let's say `startWithSystemdTarget` instead solves both of these issues.

Reply:
You could do that but I don't see how that would improve composability at all.

The only difference would be that `startWithSystemdTarget` is quite a bit more low-level while `startWithGraphical` is more abstract. The former requires knowledge about systemd while the latter just requires the user to know whether they'll use their machine with a graphical interface or not.

I usually prefer creating abstract options because the whole point of NixOS is to abstract things away. Lower-level options are still available for those who want to customise deeply; they can disable this option and set up the unit/service config any way they like.

> Imprecise naming: This option doesn't actually mean "start with graphical". It means start only with graphical.

If we used precise names for everything, we'd get enterprise-ready names and I don't think anyone wants that.

I think the description provides enough background to communicate that fact but I'm not opposed to making it more exhaustive.